### PR TITLE
Feature/embedded api

### LINKED ISSUE #19: Embedded broker as maven dependency
## Problem or use case
I would like to test my mqtt client/broker setup using JUnit tests.
Since it is cumbersome to start a broker before running the tests, an embedded MQTT 5 compliant broker would ease up the testing.

## Preferred solution or suggestions
* HiveMQ broker as maven dependency
* Documentation on how to start the broker within a running Java process (e.g. Junit test)

### PR DESCRIPTION
**Pull request by sauroter**
**Motivation**

Resolves #19 

**Changes**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -72,6 +72,7 @@ project.ext {
 
 configurations {
     provided
+    packagingOnly
```
Comment (SgtSilvio) [inline build.gradle:75]: This is not used anywhere

File: build.gradle [hunk 2]
```diff
@@ -519,12 +520,32 @@ forbiddenApisTest.enabled = false
  * Maven Publish *
  *****************/
 
+void addManifest(Jar jar) {
+    jar.manifest {
+        attributes 'Implementation-Title': project.name
+        attributes 'Implementation-Vendor': 'HiveMQ'
+        attributes 'Implementation-Version': project.version
```
Comment (SgtSilvio) [inline build.gradle:527]: We already have a manifest for the jar and the contents are different (e.g. title is HiveMQ)
We should use something like
```
tasks.withType(Jar) {
    manifest {
        attributes ...
    }
}
```

File: build.gradle [hunk 3]
```diff
@@ -519,12 +520,32 @@ forbiddenApisTest.enabled = false
...
+    }
+}
+
+task embeddedJar(type: Jar) {
```
Comment (SgtSilvio) [inline build.gradle:531]: why is this needed, we already have a jar task that does the same

File: build.gradle [hunk 4]
```diff
@@ -519,12 +520,32 @@ forbiddenApisTest.enabled = false
...
+    group 'build'
+
+    from sourceSets.main.output
+    addManifest(it)
+    exclude('META-INF/INDEX.LIST', 'META-INF/*.SF', 'META-INF/*.DSA', 'META-INF/*.RSA')
```
Comment (SgtSilvio) [inline build.gradle:536]: why is this needed, we do not have any of these files

File: build.gradle [hunk 5]
```diff
@@ -519,12 +520,32 @@ forbiddenApisTest.enabled = false
...
+}
+
+task sourceJar(type: Jar) {
```
Comment (SgtSilvio) [inline build.gradle:539]: convention for this name is sourcesJar
We can use the following when rebasing on the latest changes which update the gradle version
```
java {
    withJavadocJar()
    withSourcesJar()
}
```

File: build.gradle [hunk 6]
```diff
@@ -519,12 +520,32 @@ forbiddenApisTest.enabled = false
...
+    from sourceSets.main.allJava
+}
+
 publishing {
     publications {
         distribution(MavenPublication) {
 
             artifact packaging
-
+            artifactId 'hivemq-community-edition'
```
Comment (SgtSilvio) [inline build.gradle:548]: not needed, as by default the project name is used

File: build.gradle [hunk 7]
```diff
@@ -539,30 +560,123 @@ publishing {
...
+                    connection = 'scm:git:git@github.com:hivemq/hivemq-community-edition.git'
+                    developerConnection = 'scm:git:git@github.com:hivemq/hivemq-community-edition.git'
+                    url = 'https://github.com/hivemq/hivemq-community-edition/tree/master'
+                }
+            }
+        }
+
+        embedded(MavenPublication) {
+            from components.java
+            artifacts = [embeddedJar]
```
Comment (SgtSilvio) [inline build.gradle:607]:
```suggestion
            artifact embeddedJar
```

File: build.gradle [hunk 8]
```diff
@@ -539,30 +560,123 @@ publishing {
...
+            artifactId 'hivemq-community-edition-embedded'
+            pom {
```
Comment (SgtSilvio) [inline build.gradle:609]: the pom must be deduplicated, I have written a small gradle plugin for that

File: build.gradle [hunk 9]
```diff
@@ -539,30 +560,123 @@ publishing {
...
+                    developerConnection = 'scm:git:git@github.com:hivemq/hivemq-community-edition.git'
+                    url = 'https://github.com/hivemq/hivemq-community-edition/tree/master'
+                }
+            }
+        }
+
+        embedded(MavenPublication) {
+            from components.java
+            artifacts = [embeddedJar]
+            artifactId 'hivemq-community-edition-embedded'
```
Comment (SgtSilvio) [inline build.gradle:608]: artifact sourcesJar is missing (not needed if we use withSourcesJar as mentioned above)

File: build.gradle [hunk 10]
```diff
@@ -539,30 +560,123 @@ publishing {
...
+        }
+
+        sources(MavenPublication) {
+            from components.java
+            artifactId 'hivemq-community-edition-embedded'
+
+            artifact sourceJar {
+                classifier "sources"
             }
         }
```
Comment (SgtSilvio) [inline build.gradle:681]: This makes no sense, the sources jar is part of the publication and has not its own publication

File: build.gradle (hunk 9 above)
Comment (SgtSilvio) [reply build.gradle:608]: also we need a javadocJar, this is a requirement for maven central

File: build.gradle [hunk 11]
```diff
@@ -539,30 +560,123 @@ publishing {
...
                     connection = 'scm:git:git@github.com:hivemq/hivemq-community-edition.git'
                     developerConnection = 'scm:git:git@github.com:hivemq/hivemq-community-edition.git'
                     url = 'https://github.com/hivemq/hivemq-community-edition/tree/master'
                 }
+
+            }
+
+            pom.withXml {
+                asNode().dependencies.dependency.each { dep ->
+                    if (dep.artifactId.last().value().last() in ["logback-classic", "slf4j-log4j12"]) {
```
Comment (SgtSilvio) [inline build.gradle:667]: we should use the java-library plugin and use the right dependency scopes
hacking the pom manually is error-prone

File: src/main/java/com/hivemq/HiveMQServer.java [hunk 12]
```diff
@@ -19,6 +19,7 @@
 import com.codahale.metrics.MetricRegistry;
 import com.google.common.util.concurrent.ListenableFuture;
 import com.google.inject.Injector;
+import com.hivemq.configuration.service.entity.Listener;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/HiveMQServer.java:22]: the added dependencies are probably not needed

File: src/main/java/com/hivemq/configuration/info/SystemInformationImpl.java [hunk 13]
```diff
@@ -167,7 +182,9 @@ private File findAbsoluteAndRelative(final String fileLocation) {
         }
     }
 
-    private File getFolderWithOptionalSystemPropertyOrEnvironmentVariable(final String propertyName, final String variableName, final String defaultFolder, final boolean createFolderIfNotExists) {
+    private File setUpHiveMQFolder(
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/configuration/info/SystemInformationImpl.java:185]: missing annotations

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQ.java [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+import java.util.concurrent.CompletableFuture;
+
+@DoNotImplement
+@ThreadSafe
+public interface EmbeddedHiveMQ {
+
+    /**
+     * Start an EmbeddedHiveMQ.
+     * <p>
+     * This method can be called multiple times.
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQ.java:33]: this javadoc is unclear, does this mean that I can start, stop, start or start, start (noop)?

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQ.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+     *
+     * @return a {@link CompletableFuture} that completes when HiveMQ is started and ready
+     */
+    @NotNull CompletableFuture<Void> start();
+
+    /**
+     * Stop an EmbeddedHiveMQ. Calling stop clears the metric registry returned by {@link #getMetricRegistry()}.
+     * <p>
+     * This method can be called multiple times.
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQ.java:42]: same here

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.nio.file.Path;
+
+@DoNotImplement
+public interface EmbeddedHiveMQBuilder {
+
+    /**
+     * @return a new EmbeddedHiveMQBuilder.
+     */
+    static @NotNull EmbeddedHiveMQBuilder newBuilder() {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:32]: why not just builder?

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+     * @see <a href="https://www.hivemq.com/docs/hivemq/latest/user-guide/configuration.html#folders">HiveMQ
+     *         Folders</a>
+     */
+    @NotNull EmbeddedHiveMQBuilder withConfigurationFolder(@NotNull Path configFolder);
+
+    /**
+     * This sets the HIVEMQ_DATA_FOLDER to the given argument. Using this method overrides all other ways to configure
+     * it.
+     *
+     * @param dataFolder the used configuration folder.
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:51]:
```suggestion
     * @param dataFolder the used data folder.
```

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * This sets the HIVEMQ_CONFIG_FOLDER to the given argument. Using this method overrides all other ways to configure
+     * it.
+     *
+     * @param configFolder the used configuration folder.
+     * @return this builder.
+     * @see <a href="https://www.hivemq.com/docs/hivemq/latest/user-guide/configuration.html#folders">HiveMQ
+     *         Folders</a>
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:43]:
```suggestion
     *         folders</a>
```

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+    @NotNull EmbeddedHiveMQBuilder withConfigurationFolder(@NotNull Path configFolder);
+
+    /**
+     * This sets the HIVEMQ_DATA_FOLDER to the given argument. Using this method overrides all other ways to configure
+     * it.
+     *
+     * @param dataFolder the used configuration folder.
+     * @return this builder.
+     * @see <a href="https://www.hivemq.com/docs/hivemq/latest/user-guide/configuration.html#folders">HiveMQ
+     *         Folders</a>
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:54]:
```suggestion
     *         folders</a>
```

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+    @NotNull EmbeddedHiveMQBuilder withDataFolder(@Nullable Path dataFolder);
+
+    /**
+     * This sets the HIVEMQ_EXTENSION_FOLDER to the given argument. Using this method overrides all other ways to
+     * configure it.
+     *
+     * @param extensionsFolder the used extensions folder.
+     * @return this builder.
+     * @see <a href="https://www.hivemq.com/docs/hivemq/latest/user-guide/configuration.html#folders">HiveMQ
+     *         Folders</a>
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:65]:
```suggestion
     *         folders</a>
```

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    @NotNull EmbeddedHiveMQBuilder withExtensionFolder(@NotNull Path extensionsFolder);
+
+    /**
+     * @return a new EmbeddedHiveQM with the configured HiveMQ folders.
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:70]:
```suggestion
     * @return a new EmbeddedHiveMQ with the configured HiveMQ folders.
```

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.io.File;
+import java.nio.file.Path;
+
+/**
+ * @author Georg Held
+ */
+public class EmbeddedHiveMQBuilderImpl implements EmbeddedHiveMQBuilder {
+
+    private @Nullable Path conf = null;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java:32]: I would name the attributes explicitly (configFolder), conf can be misunderstood as the path to the config file.
Same for the folders below

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * This sets the HIVEMQ_DATA_FOLDER to the given argument. Using this method overrides all other ways to configure
+     * it.
+     *
+     * @param dataFolder the used configuration folder.
+     * @return this builder.
+     * @see <a href="https://www.hivemq.com/docs/hivemq/latest/user-guide/configuration.html#folders">HiveMQ
+     *         Folders</a>
+     */
+    @NotNull EmbeddedHiveMQBuilder withDataFolder(@Nullable Path dataFolder);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:56]: why is this argument nullable and the others notnull?

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * This sets the HIVEMQ_EXTENSION_FOLDER to the given argument. Using this method overrides all other ways to
+     * configure it.
+     *
+     * @param extensionsFolder the used extensions folder.
+     * @return this builder.
+     * @see <a href="https://www.hivemq.com/docs/hivemq/latest/user-guide/configuration.html#folders">HiveMQ
+     *         Folders</a>
+     */
+    @NotNull EmbeddedHiveMQBuilder withExtensionFolder(@NotNull Path extensionsFolder);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:67]: imho this should be withExtensionsFolder, otherwise I would think this is the folder of a single extension

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+    private @Nullable Path data = null;
+    private @Nullable Path extensions = null;
+
+    @Override
+    public @NotNull EmbeddedHiveMQBuilder withConfigurationFolder(final @Nullable Path configFolder) {
+        conf = configFolder;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java:38]: Where is the null check? the interface defines the parameter as notnull
either change the interface to nullable or add a null check here

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 26]
```diff
@@ -0,0 +1,203 @@
...
+        if ((currentState == State.STOPPED && newState == State.NOT_STARTED)
+                || newState.isHigher(currentState)) {
+            log.debug("Advancing EmbeddedHiveMQState from \"{}\" to \"{}\"", currentState, newState);
+            currentState = newState;
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> start() {
+        advanceState(State.NOT_STARTED);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:94]: why is this necessary? the sate is already NOT_STARTED, if it is not -> else

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+                final HiveMQServer hiveMQServer = injector.getInstance(HiveMQServer.class);
+                try {
+                    hiveMQServer.start();
+                } catch (final Exception e) {
+                    throw new HiveMQServerException(e);
+                }
+                advanceState(State.RUNNING);
+            }).whenComplete(HiveMQServerException.handler(startFuture));
+        } else {
+            log.warn("Could not start EmbeddedHiveMQ. Reason: \"Already started\"");
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:114]: imho this is weird, if we call start after stop we will only get a log, but the old startFuture which will return success.
I think this case should return an errored future.
If we want to expose the startFuture, we should add an extra method with clear naming (like startedFuture) and clear documentation. "start" includes the actual action to start HiveMQ.

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 28]
```diff
@@ -0,0 +1,203 @@
...
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> start() {
+        advanceState(State.NOT_STARTED);
+        if (currentState == State.NOT_STARTED) {
+            advanceState(State.STARTING);
+            log.info("Starting EmbeddedHiveMQ.");
+
+            startFuture = new CompletableFuture<>();
+            CompletableFuture.runAsync(() -> {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:100]: I don't think it is a good idea to use a shared executor for the startup, because this blocks for some time.
I think a new thread here would be the best solution

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+                    hiveMQShutdownHook.run();
+                }
+
+                shutdownHooks.clearRuntime();
+                metricRegistry.removeMatching(MetricFilter.ALL);
+                injector = null;
+                advanceState(State.STOPPED);
+            });
+        } else {
+            log.warn("Could not stop EmbeddedHiveMQ. Reason: \"Already stopping\"");
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:168]: imho this should also return an exceptional future

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 30]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    @Override
+    public synchronized @NotNull CompletableFuture<Void> stop() {
+        if (currentState == State.NOT_STARTED) {
+            log.warn("Could not stop EmbeddedHiveMQ. Reason: \"Not started\"");
+            stopFuture = CompletableFuture.completedFuture(null);
+
+        } else if (currentState.isLower(State.STOPPING)) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:144]: does it work when calling stop before the start future is finished?

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 31]
```diff
@@ -0,0 +1,203 @@
...
+            advanceState(State.STOPPING);
+            log.info("Stopping EmbeddedHiveMQ.");
+
+            stopFuture = startFuture.thenRunAsync(() -> {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:148]: I don't think it is a good idea to use a shared executor for stop as well, because this blocks for some time.
I think a new thread here would be the best solution

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java [hunk 32]
```diff
@@ -0,0 +1,203 @@
...
+        }
+        return stopFuture;
+    }
+
+    @Override
+    public @NotNull MetricRegistry getMetricRegistry() {
+        return metricRegistry;
+    }
+
+    private static class HiveMQServerException extends RuntimeException {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:178]: this workaround would also be not needed when using a new thread for start and stop and use try catch all

File: src/main/java/com/hivemq/persistence/SingleWriterService.java [hunk 33]
```diff
@@ -208,15 +213,34 @@ public AtomicInteger getRunningThreadsCount() {
...
-    }
-
     @NotNull
     public ExecutorService[] getCallbackExecutors() {
         return callbackExecutors;
     }
 
+    public void stop() {
+        final long start = System.currentTimeMillis();
+        if (log.isTraceEnabled()) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/SingleWriterService.java:223]: this if is then unnecessary

File: src/main/java/com/hivemq/persistence/SingleWriterService.java [hunk 34]
```diff
@@ -208,15 +213,34 @@ public AtomicInteger getRunningThreadsCount() {
...
+        try {
+            singleWriterExecutor.awaitTermination(shutdownGracePeriod, TimeUnit.SECONDS);
+            if (log.isTraceEnabled()) {
+                log.trace("Finished single writer shutdown in {} ms", (System.currentTimeMillis() - start));
+            }
+        } catch (final InterruptedException e) {
+            //ignore
+        }
+        singleWriterExecutor.shutdownNow();
+        for (final @NotNull ExecutorService callbackExecutor : callbackExecutors) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/SingleWriterService.java:238]:
```suggestion
        for (final ExecutorService callbackExecutor : callbackExecutors) {
```

File: src/main/java/com/hivemq/persistence/PersistenceShutdownHook.java [hunk 35]
```diff
@@ -101,9 +101,8 @@ public boolean isAsynchronous() {
 
     @Override
     public void run() {
-        long start = 0;
+        final long start = System.currentTimeMillis();
         if (log.isTraceEnabled()) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/PersistenceShutdownHook.java:105]: this if is then unnecessary

File: build.gradle (hunk 5 above)
Comment (sauroter) [reply build.gradle:539]: nice 👍

File: build.gradle (hunk 4 above)
Comment (sauroter) [reply build.gradle:536]: you are correct, old habbit.

File: build.gradle (hunk 3 above)
Comment (sauroter) [reply build.gradle:531]: you mean jar?

File: build.gradle (hunk 1 above)
Comment (sauroter) [reply build.gradle:75]: i guess provided is neither...

File: build.gradle (hunk 6 above)
Comment (sauroter) [reply build.gradle:548]: ok

File: build.gradle (hunk 8 above)
Comment (sauroter) [reply build.gradle:609]: Can you specify please.

File: build.gradle (hunk 9 above)
Comment (sauroter) [reply build.gradle:608]: 👍

File: build.gradle (hunk 10 above)
Comment (sauroter) [reply build.gradle:681]: 👍

File: build.gradle (hunk 1 above)
Comment (SgtSilvio) [reply build.gradle]: yes, you just have to rebase, then provided is removed

File: build.gradle (hunk 3 above)
Comment (SgtSilvio) [reply build.gradle]: yes, jar packages only our classes (shadowJar packages also all dependencies)

File: src/main/java/com/hivemq/HiveMQServer.java (hunk 12 above)
Comment (sauroter) [reply src/main/java/com/hivemq/HiveMQServer.java]: Done

File: src/main/java/com/hivemq/configuration/info/SystemInformationImpl.java (hunk 13 above)
Comment (sauroter) [reply src/main/java/com/hivemq/configuration/info/SystemInformationImpl.java]: done

File: build.gradle (hunk 8 above)
Comment (SgtSilvio) [reply build.gradle:529]: https://github.com/hivemq/hivemq-extension-sdk/pull/3/files#diff-392475fdf2bc320d17762ed97109a121R17

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java (hunk 22 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java]: done

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java (hunk 24 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java]: yes

File: src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java (hunk 23 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/EmbeddedHiveMQBuilder.java:56]: all should be nullable

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java (hunk 25 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQBuilderImpl.java]: none needed, see above

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 26 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:90]: For the restart case

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 30 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: yes

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 28 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:96]: I dont like spawning a new thread per Embedded HiveMQ. What exactly is the advantage for this? Now it is configurable for the user. ut perhapps we should make this configuration explicit? What do you think about allowing to set an executor in the builder?

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 29 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: i don't think so as this would break idempotency.

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 31 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: see answer above

File: src/main/java/com/hivemq/persistence/SingleWriterService.java (hunk 33 above)
Comment (sauroter) [reply src/main/java/com/hivemq/persistence/SingleWriterService.java:223]: This was just moved. I do not think we should differ the code bases to much.

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 27 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: hm, I agree. The stop now works while the start is still in progress but not vice versa. Would it be good if both could be triggered while the other is in progress?
Like the *level based* system of kubernetes?

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 28 above)
Comment (SgtSilvio) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java:96]: HiveMQ takes a long time to start (even when we optimize it to under 1sec this is still very long). Blocking a shared executor for such a long time seems wrong to me, we actually stall other work that is totally unrelated during the HiveMQ startup/shutdown.
I would not add a configuration for that right now in the builder, as the right way is to make the start of HiveMQ return a Future anyway.

**Comment by LBrandl** [comment]
Looks like the "extension-start-stop-executor" is not stop and some Xodus threads are also still running when stop is called.
- Exodus shared timer 
- Exodus shared background cleaner
- Exodus shared deferred I/O job processor
- threadJobProcessorPoolSpawner

**Comment by sauroter** [comment]
@LBrandl 
I think these are all Daemon Threads but I will check again.

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 27 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: I implemented this.

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 28 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: Done

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 29 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: Found a new solution for this.

File: src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java (hunk 32 above)
Comment (sauroter) [reply src/main/java/com/hivemq/embedded/internal/EmbeddedHiveMQImpl.java]: Was implemented as suggested.

File: build.gradle (hunk 8 above)
Comment (sauroter) [reply build.gradle]: done

File: build.gradle (hunk 11 above)
Comment (sauroter) [reply build.gradle]: I am not sure if this is solved. Please take a look again.

File: build.gradle [hunk 36]
```diff
@@ -23,24 +25,58 @@ description = 'HiveMQ CE is a Java-based open source MQTT broker that fully supp
...
+            it.id = "SgtSilvio"
+            it.name = "Silvio Giebl"
+            it.email = "silvio.giebl@hivemq.com"
+        }
+    }
+    github {
+        org = "hivemq"
+        repo = "hivemq-community-edition"
+        issues()
+        pages()
```
Comment (SgtSilvio) [inline build.gradle:68]: We do not have GitHub Pages for HiveMQ CE

Comment (sauroter) [reply build.gradle:68]: Didnt you create one?

Comment (sauroter) [reply build.gradle:68]: Noe I am confused

Comment (SgtSilvio) [reply build.gradle:68]: We do not have "https://hivemq.github.io/hivemq-community-edition/"
We just have "https://hivemq.github.io/nightly-builds/" but this is something different

Comment (SgtSilvio) [inline build.gradle:68]:
```suggestion
```

Comment (sauroter) [reply build.gradle]: ok

File: build.gradle (hunk 11 above)
Comment (SgtSilvio) [reply build.gradle]: Actually we could use optional features with their own dependency configurations. Then the dependencies would be listed as optional in the pom

Comment (SgtSilvio) [reply build.gradle]: But I do not think this should delay this ticket.

File: src/main/java/com/hivemq/configuration/info/SystemInformation.java [hunk 37]
```diff
@@ -75,4 +75,9 @@
      * @return the count of CPUs HiveMQ uses
      */
     int getProcessorCount();
+
+    /**
+     * @return is HiveMQ running in embedded mode
```
Comment (brickhof) [inline src/main/java/com/hivemq/configuration/info/SystemInformation.java:80]: Does this mean: * @return 'whether HiveMQ is running in embedded mode'?

Comment (sauroter) [reply src/main/java/com/hivemq/configuration/info/SystemInformation.java:80]: yes, but the *is* is a coding convention. All booleans are *is*

File: src/main/java/com/hivemq/common/shutdown/ShutdownHooks.java [hunk 38]
```diff
@@ -174,8 +177,8 @@ public synchronized void remove(@NotNull final HiveMQShutdownHook hiveMQShutdown
     }
 
     /**
-     * Provides a Set of all asynchronous shutdown hooks. Async Shutdown Hooks are already registered to the JVM,
-     * so this Set is read only
+     * Provides a Set of all asynchronous shutdown hooks. Async Shutdown Hooks are already registered to the JVM, so
```
[the remark this answers is not in the document]
Comment (sauroter) [reply src/main/java/com/hivemq/common/shutdown/ShutdownHooks.java:180]: Actually big Set is right, as it refers to a data structure
